File: backend/app/schemas/content_location.py
# =============================================================================
# File: content_location.py
# Module/Service: Citation Verification / Search Service
# Layer: Schema
# Purpose: Shared ContentLocation + Citation/Search schemas for FR5 locators.
# Responsibilities:
#   - Expose page_number XOR section_index per file_type (OpenAPI ContentLocation)
#   - Helpers to build location payloads from document_chunks
# Dependencies:
#   - Pydantic; app.models.knowledge.DocumentChunk (for helper typing)
# Public Exports:
#   - ContentLocation, CitationResponse, SearchResultItem, SearchResultResponse
#   - content_location_from_chunk
# Database/Table: document_chunks (page_number, section_index, section)
# Related Modules: Enterprise_notebooklm_openapi.yaml; FR5 Business Context
# Important Notes: Do not invent the missing locator field from the other one.
# =============================================================================

from __future__ import annotations

from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field


class ContentLocation(BaseModel):
    """Vị trí nội dung trong tài liệu gốc (đúng 1 locator theo file_type)."""

    page_number: int | None = Field(
        default=None,
        description="Physical page/slide/sheet (PDF/PPTX/XLSX). Null for DOCX.",
    )
    section_index: int | None = Field(
        default=None,
        description="1-based logical section index (DOCX). Null for PDF/PPTX/XLSX.",
    )
    section_title: str | None = Field(
        default=None,
        description="Nearest heading/sheet/slide title; used especially for DOCX UI.",
    )


class CitationResponse(BaseModel):
    id: UUID
    message_id: UUID
    retrieval_id: UUID
    document_id: UUID
    text_snippet: str
    verified: bool
    order_index: int
    location: ContentLocation | None = None


class SearchResultItem(BaseModel):
    chunk_id: UUID | None = None
    entity_id: UUID | None = None
    document_id: UUID
    text_snippet: str
    retrieval_method: Literal["vector", "bm25", "knowledge_graph", "rerank"]
    score: float
    rank: int
    location: ContentLocation | None = None


class SearchResultResponse(BaseModel):
    results_count: int
    results: list[SearchResultItem] = Field(default_factory=list)


def content_location_from_chunk(
    *,
    page_number: int | None,
    section_index: int | None,
    section: str | None = None,
) -> ContentLocation:
    """Build API location from chunk columns without inventing missing fields."""
    return ContentLocation(
        page_number=page_number,
        section_index=section_index,
        section_title=section,
    )


def content_location_from_mapping(data: dict[str, Any]) -> ContentLocation:
    """Build location from chunk ORM/dict/payload fields."""
    return content_location_from_chunk(
        page_number=data.get("page_number"),
        section_index=data.get("section_index"),
        section=data.get("section") or data.get("section_title"),
    )

File: frontend/lib/content-location.ts
/**
 * =============================================================================
 * File: content-location.ts
 * Module/Service: Citation / Documents UI
 * Layer: UI
 * Purpose: Format ContentLocation + document extent labels (FR5 convention).
 * Responsibilities:
 *   - PDF/PPTX/XLSX → "Trang X"; DOCX → "Mục X" (+ section title)
 *   - Document list page_count: DOCX → "X mục", others → "X trang"
 * Dependencies:
 *   - OpenAPI ContentLocation; Business Context FR5
 * Public Exports:
 *   - ContentLocation, formatContentLocationLabel, formatDocumentExtentLabel
 * Database/Table: N/A
 * Related Modules: features/citation/CitationLocationLabel
 * Important Notes: Never show "Trang X" for DOCX section_index.
 * =============================================================================
 */

export type ContentLocation = {
  page_number?: number | null;
  section_index?: number | null;
  section_title?: string | null;
};

export type DocumentFileType = "pdf" | "docx" | "xlsx" | "pptx" | "txt";

/** Citation chip label; null when no locator (hide position UI). */
export function formatContentLocationLabel(
  location: ContentLocation | null | undefined,
): string | null {
  if (!location) return null;
  if (location.page_number != null) {
    return `Trang ${location.page_number}`;
  }
  if (location.section_index != null) {
    const title = (location.section_title || "").trim();
    return title
      ? `Mục ${location.section_index}: ${title}`
      : `Mục ${location.section_index}`;
  }
  return null;
}

/** Document list extent from page_count + file_type. */
export function formatDocumentExtentLabel(
  fileType: DocumentFileType | string | null | undefined,
  pageCount: number | null | undefined,
): string | null {
  if (pageCount == null || pageCount < 0) return null;
  if (fileType === "docx") {
    return pageCount === 1 ? "1 mục" : `${pageCount} mục`;
  }
  if (fileType === "pptx") {
    return pageCount === 1 ? "1 slide" : `${pageCount} slide`;
  }
  if (fileType === "xlsx") {
    return pageCount === 1 ? "1 sheet" : `${pageCount} sheet`;
  }
  return pageCount === 1 ? "1 trang" : `${pageCount} trang`;
}
